### PR TITLE
change farming code and reduce farming texture sizes (adv- and optipng)

### PR DESCRIPTION
**Pull request by HybridDog**

### CODE REVIEW

File: mods/farming/API.txt [hunk 1]
```diff
@@ -25,4 +20,4 @@ Plant definition
 	^ Always provide a plant texture for each step, format: modname_plantname_i.png (i = stepnumber)
 	minlight = 13, -- Minimum light to grow
 	maxlight = default.LIGHT_MAX -- Maximum light to grow
-}
\ No newline at end of file
+}
```
Comment (HybridDog) [inline mods/farming/API.txt]: What did geany change here?

File: mods/farming/nodes.lua [hunk 2]
```diff
@@ -89,64 +89,78 @@ minetest.register_abm({
...
+local defchange = {drop = {
+	max_items = 1,
+	items = {
+		{items = {'farming:seed_wheat'},rarity = 5},
+		{items = {'default:grass_1'}},
+	}
+}}
+for i = 1, 5 do
+	minetest.override_item("default:grass_"..i, defchange)
 end
```
Comment (HybridDog) [inline mods/farming/nodes.lua]: ↑ this should make it a bit faster

File: mods/farming/api.lua [hunk 3]
```diff
@@ -1,77 +1,64 @@
...
+	if not pt
+	or pt.type ~= "node"
+	or pt.above.y ~= pt.under.y+1 then
 		return
 	end
-	if pt.type ~= "node" then
+
+	local above = minetest.get_node(pt.above)
+	if above.name ~= "air"
+	or not minetest.registered_nodes[above.name] then
```
Comment (kilbith) [inline mods/farming/api.lua]: Move `or not` on the line above and indent twice times the line below.

**Comment by kilbith** [comment]
Please read the [Lua code style guidelines](http://dev.minetest.net/Lua_code_style_guidelines).

**Comment by HybridDog** [comment]
>Please read the Lua code style guidelines.

sorry, l thought they would only refer to builtin

File: mods/farming/api.lua [hunk 4]
```diff
@@ -184,8 +145,11 @@ farming.register_plant = function(name, def)
 	for k, v in pairs(def.fertility) do
 		g[v] = 1
 	end
+
+	local mname, pname = unpack(name:split(":"))
+
 	minetest.register_node(":" .. mname .. ":seed_" .. pname, {
-		description = def.description,
+		description = def.description or "Seed",
```
Comment (rubenwardy) [inline mods/farming/api.lua]: This is a bad idea, as it makes less useful descriptions of items - may just require description, and fail gracefully:

```Lua
if not def.description then
	error("Description needed in farming definition!")
end
```

File: mods/farming/api.lua [hunk 5]
```diff
@@ -213,16 +178,15 @@ farming.register_plant = function(name, def)
 
 	-- Register growing steps
 	for i=1,def.steps do
```
Comment (rubenwardy) [inline mods/farming/api.lua]: You should fix the code around your changes, as you're touching most of the file.

Comment (rubenwardy) [inline mods/farming/api.lua]: You should fix the code around your changes, as you're touching most of the file.

Comment (rubenwardy) [inline mods/farming/api.lua]: You should fix the code around your changes, as you're touching most of the file.

File: mods/farming/api.lua [hunk 6]
```diff
@@ -240,6 +211,13 @@ farming.register_plant = function(name, def)
 		})
 	end
 
+	if not def.minlight then
```
Comment (rubenwardy) [inline mods/farming/api.lua]:
```Lua
def.minlight = def.minlight or 1
```
?

File: mods/farming/api.lua [hunk 7]
```diff
@@ -247,40 +225,42 @@ farming.register_plant = function(name, def)
...
+				for _, v in pairs(fertility) do
 					if minetest.get_item_group(soil_node.name, v) ~= 0 then
 						can_grow = true
+						break
 					end
 				end
-				if can_grow then
-					minetest.set_node(pos, {name = node.name:gsub("seed_", "") .. "_1"})
+				if not can_grow then
+					return
```
Comment (rubenwardy) [inline mods/farming/api.lua]: this change makes no sense.

```Lua
if can_grow then
    node.name = node.name:gsub("seed_", "") .. "_1"
    minetest.set_node(pos, node)
end
return
```

File: mods/farming/nodes.lua [hunk 8]
```diff
@@ -54,7 +70,15 @@ minetest.register_node("farming:desert_sand_soil", {
...
 	tiles = {"farming_desert_sand_soil.png", "default_desert_sand.png"},
-	groups = {crumbly=3, not_in_creative_inventory = 1, falling_node=1, sand=1, soil = 2, desert = 1, field = 1},
+	groups = {
+		crumbly=3,
+		not_in_creative_inventory = 1,
+		falling_node=1,
+		sand=1,
+		soil = 2,
+		desert = 1,
+		field = 1
```
Comment (rubenwardy) [inline mods/farming/nodes.lua]: inconsistent ` = `

File: mods/farming/api.lua [hunk 9]
```diff
@@ -1,77 +1,59 @@
...
 -- TODO Ignore group:flower
-farming.hoe_on_use = function(itemstack, user, pointed_thing, uses)
-	local pt = pointed_thing
-	-- check if pointing at a node
-	if not pt then
+
+local creative = minetest.setting_getbool("creative_mode")
+function farming.hoe_on_use(itemstack, user, pt, uses)
+	-- check if pointing at a node's top
+	if not pt or pt.type ~= "node" or pt.above.y ~= pt.under.y+1 then
```
Comment (rubenwardy) [inline mods/farming/api.lua]: maybe

```Lua
if not (pt and pt.type == "node" and pt.above.y == pt.under.y + 1) then
```

**Comment by HybridDog** [comment]
How about now?

File: mods/farming/API.txt (hunk 1 above)
Comment (nerzhul) [inline mods/farming/API.txt]: New line

File: mods/farming/api.lua [hunk 10]
```diff
@@ -1,77 +1,59 @@
...
 	})
-	
-	if not minetest.setting_getbool("creative_mode") then
-		itemstack:add_wear(65535/(uses-1))
+
+	if creative then
+		return
 	end
+
+	itemstack:add_wear(65535/(uses-1))
```
Comment (rubenwardy) [inline mods/farming/api.lua]: can uses either equals 1?

File: mods/farming/api.lua [hunk 11]
```diff
@@ -212,17 +177,16 @@ farming.register_plant = function(name, def)
...
 	-- Register growing steps
-	for i=1,def.steps do
-		local drop = {
-			items = {
-				{items = {mname .. ":" .. pname}, rarity = 9 - i},
-				{items = {mname .. ":" .. pname}, rarity= 18 - i * 2},
-				{items = {mname .. ":seed_" .. pname}, rarity = 9 - i},
-				{items = {mname .. ":seed_" .. pname}, rarity = 18 - i * 2},
-			}
+	for i = 1,def.steps do
```
Comment (rubenwardy) [inline mods/farming/api.lua]: space after `,`

File: mods/farming/api.lua [hunk 12]
```diff
@@ -247,40 +221,37 @@ farming.register_plant = function(name, def)
...
 			-- grow seed
-			if minetest.get_item_group(node.name, "seed") and node_def.fertility then
-				local can_grow = false
-				local soil_node = minetest.get_node_or_nil({x = pos.x, y = pos.y - 1, z = pos.z})
-				if not soil_node then
-					return
-				end
-				for _, v in pairs(node_def.fertility) do
+			if fertility
+			and minetest.get_item_group(node.name, "seed") then
```
Comment (rubenwardy) [inline mods/farming/api.lua]: indent twice

**Comment by rubenwardy** [comment]
looks good, apart from the mentioned issues

File: mods/farming/API.txt (hunk 1 above)
Comment (HybridDog) [inline mods/farming/API.txt]: ok, l thought github would show a red arrow if a line was added (or removed)

File: mods/farming/api.lua (hunk 10 above)
Comment (HybridDog) [inline mods/farming/api.lua]: it would abort with a "hoe max uses are invalid" error message

**Comment by HybridDog** [comment]
>looks good, apart from the mentioned issues

fixed

File: mods/farming/api.lua (hunk 10 above)
Comment (rubenwardy) [inline mods/farming/api.lua]: `max_uses` would be a better variable name, here.

Comment (HybridDog) [inline mods/farming/api.lua]: fixed

**Comment by rubenwardy** [comment]
Why the close?

**Comment by HybridDog** [comment]
>Why the close?

When loading this page, firefox is lagging a lot.

**Pull request closed without merging**